aip/client-libraries/4234.md

```diff
@@ -0,0 +1,92 @@
+---
+id: 4234
+state: draft
+created: 2021-02-10
+---
+
+# Common service client mix-ins
+
+An API platform often has utility services that are shared amongst the discrete
+product services. These common services each have a centrally-defined surface,
+but individual instances are hosted alongside each product service. Thus, the
+surfaces of each common services hosted by the product service are "mixed in" at
+runtime.
+
+
+## Guidance
+
+Client libraries **may** provide client methods for the common services declared
+to be mixed-in by a product service. Such methods improve the user experience by
+presenting the mix-in methods from a client configured to communicate with the
+product service that hosts the mix-in service.
+
+Should client libraries support mix-in services, the common services that
+**must** be supported include:
+
+- `google.cloud.location.Locations`
+- `google.iam.v1.IAMPolicy`
+- `google.longrunning.Operations`
+
+**Note:** The list of supported mix-in services and the mix-in services
+themselves change infrequently. Any new common service to be supported should be
+added to the list here and generators should be updated.
+
+A mix-in service **must** be declared in the [`apis`] in a
+[`google.api.Service`] to be generated into a client library. Furthermore, it
+**must** have [`google.api.Http`] bindings declared for each of its RPC's in the
+[`http`] configuration of the same [`google.api.Service`] (more details on these
+bindings in [AIP-127]).
+
+
+### Implementing mix-in support
+
+#### Generator configuration
+
+Client library generators **must** accept the file path of a
+[`google.api.Service`] in YAML form. This flag **must** be optional - meaning
+the flag **must not** be required for basic client generation that omits the
+mix-in methods.
+
+#### Mix-in API client configuration
+
+If a mix-in API utilizes client library configuration such as
+`google.api.method_signature` or request header injection, generators **may**
+support it in the generated mix-in methods if the configuration is accessible.
+
+#### Multiple host service clients
+
+If there are multiple services defined by the host API that would result in
+multiple clients being generated, generators **should** include the mix-in
+methods on all eligible, generated service clients.
+
+#### Overriding a duplicate RPC
+
+Client library generators **must not** generate a mix-in method on a host
+service client if that service already defines an RPC with the same name. For
```

Comment:
For learning purpose, as client library generator, how to detect if the method is mix-in method? 
My guess is that run through the existing  mix-in service and dive in each method to see if any match.

Reply:
> run through the existing mix-in service and dive in each method to see if any match.

Yep, pretty much! Basically `if contains(service.GetMethods(), "GetIamPolicy") { skipMixinMethod()}`